File: gateco/apps/frontend/src/types/identity-provider.ts
export type IdentityProviderType = 'azure_entra_id' | 'aws_iam' | 'gcp' | 'okta';

export type IdentityProviderStatus = 'connected' | 'error' | 'syncing' | 'disconnected';

export interface SyncConfig {
  auto_sync: boolean;
  sync_interval_minutes: number;
  sync_groups: boolean;
  sync_roles: boolean;
  sync_attributes: boolean;
}

export interface IdentityProvider {
  id: string;
  name: string;
  type: IdentityProviderType;
  status: IdentityProviderStatus;
  sync_config: SyncConfig;
  principal_count: number;
  group_count: number;
  last_sync: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateIdentityProviderRequest {
  name: string;
  type: IdentityProviderType;
  config: Record<string, string>;
  sync_config: SyncConfig;
}

File: gateco/apps/backend/src/gateco/database/models/identity_provider.py
"""IdentityProvider model — external identity sources."""

import datetime
from typing import Any, Dict, Optional

from sqlalchemy import DateTime, Enum, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from gateco.database.enums import IdentityProviderStatus, IdentityProviderType
from gateco.database.models.base import (
    Base,
    OrganizationScopedMixin,
    SoftDeleteMixin,
    TimestampMixin,
    UUIDPrimaryKeyMixin,
)


class IdentityProvider(
    Base, UUIDPrimaryKeyMixin, OrganizationScopedMixin, TimestampMixin, SoftDeleteMixin
):
    """External identity provider connection."""

    __tablename__ = "identity_providers"

    name: Mapped[str] = mapped_column(String(200), nullable=False)
    type: Mapped[IdentityProviderType] = mapped_column(
        Enum(IdentityProviderType, native_enum=True, name="identity_provider_type"),
        nullable=False,
    )
    status: Mapped[IdentityProviderStatus] = mapped_column(
        Enum(IdentityProviderStatus, native_enum=True, name="identity_provider_status"),
        nullable=False,
        default=IdentityProviderStatus.disconnected,
    )
    config: Mapped[Optional[Dict[str, Any]]] = mapped_column(JSONB, nullable=True)
    sync_config: Mapped[Optional[Dict[str, Any]]] = mapped_column(JSONB, nullable=True)
    principal_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    group_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    last_sync: Mapped[Optional[datetime.datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    def __repr__(self) -> str:
        return f"<IdentityProvider(id={self.id}, name={self.name!r}, type={self.type.value})>"


__all__ = ["IdentityProvider"]
